core/validate.py

```diff
@@ -20,39 +44,31 @@ def spot_exists(self, ui):
             user_shot_choice = ui.get_input('>>')
         return user_shot_choice
 
-    def spot_occupied(self, board_state, ui):
+    def spot_occupied(self, board_state, ui, all_ships):
+        ship_symbols = [] 
         user_shot_choice = self.spot_exists(ui)
         user_letter = user_shot_choice[0]
         user_num = user_shot_choice[1:]
-        lets = {
-                'A': 0,
-                'B': 1,
-                'C': 2,
-                'D': 3,
-                'E': 4,
-                'F': 5,
-                'G': 6,
-                'H': 7,
-                'I': 8,
-                'J': 9
-                }
-        nums = {
-                '1': 0,
-                '2': 1,
-                '3': 2,
-                '4': 3,
-                '5': 4,
-                '6': 5,
-                '7': 6,
-                '8': 7,
-                '9': 8,
-                '10': 9
-                }
 
-        while board_state[lets[user_letter]][nums[user_num]] is not None:
+        for ship in all_ships:
+            ship_symbols.append(ship['symbol'])
+            
+        # must check for None AND any ship symbols
```

Comment:
no comments. if it's too hard to understand without the comment there might be some changes that need to be made